src/profile-logic/import/instruments/index.js

```diff
@@ -0,0 +1,910 @@
+      object.clear();
+      for (const [k, v] of clone.entries()) {
+        object.set(visit(k), visit(v));
+      }
+    }
+    return object;
+  };
+  for (let i = 0; i < root.$objects.length; i++) {
+    visit(root.$objects[i]);
+  }
+  return visit(root.$top);
+}
+
+// This function creates an archived data for given buffer by interpreting various Instruments specific data types
+function readInstrumentsArchive(buffer) {
+  const byteArray = new Uint8Array(buffer);
+  const parsedPlist = parseBinaryPlist(byteArray);
+
+  const data = expandKeyedArchive(parsedPlist, ($classname, object) => {
+    switch ($classname) {
+      case 'NSTextStorage':
+      case 'NSParagraphStyle':
+      case 'NSFont':
+        // Stuff that's irrelevant for constructing a flamegraph
+        return null;
+
+      case 'PFTSymbolData': {
+        const ret = Object.create(null);
+        ret.symbolName = object.$0;
+        ret.sourcePath = object.$1;
+        ret.addressToLine = new Map<any, any>();
+        for (let i = 3; ; i += 2) {
+          const address = object['$' + i];
+          const line = object['$' + (i + 1)];
+          // eslint-disable-next-line eqeqeq
+          if (address == null || line == null) {
+            break;
+          }
+          ret.addressToLine.set(address, line);
+        }
+        return ret;
+      }
+
+      case 'PFTOwnerData': {
+        const ret = Object.create(null);
+        ret.ownerName = object.$0;
+        ret.ownerPath = object.$1;
+        return ret;
+      }
+
+      case 'PFTPersistentSymbols': {
+        const ret = Object.create(null);
+        const symbolCount = object.$4;
+
+        ret.threadNames = object.$3;
+        ret.symbols = [];
+        for (let i = 1; i < symbolCount; i++) {
+          ret.symbols.push(object['$' + (4 + i)]);
+        }
+        return ret;
+      }
+
+      case 'XRRunListData': {
+        const ret = Object.create(null);
+        ret.runNumbers = object.$0;
+        ret.runData = object.$1;
+        return ret;
+      }
+
+      case 'XRIntKeyedDictionary': {
+        const ret = new Map();
+        const size = object.$0;
+        for (let i = 0; i < size; i++) {
+          const key = object['$' + (1 + 2 * i)];
+          const value = object['$' + (1 + (2 * i + 1))];
+          ret.set(key, value);
+        }
+        return ret;
+      }
+
+      case 'XRCore': {
+        const ret = Object.create(null);
+        ret.number = object.$0;
+        ret.name = object.$1;
+        return ret;
+      }
+      default:
+    }
+    return object;
+  });
+  return data;
+}
+
+// This function adds a padding ahead of given string to make its length equal to a given width value
+function zeroPad(s: string, width: number) {
+  return new Array(Math.max(width - s.length, 0) + 1).join('0') + s;
+}
+
+function getOrThrow<K, V>(map: Map<K, V>, k: K): V {
+  if (!map.has(k)) {
+    if (typeof k === 'string') {
+      throw new Error(`Expected key ${k}`);
+    }
```

Comment:
I'd put this check at the end of the function, and remove the `if (!map.has(k)) {`, so that we don't look up the value twice.